saving/game_save_manager.py

```diff
@@ -0,0 +1,183 @@
+    sanitised_timestamp = datetime.now().isoformat(timespec='seconds').replace(':', '.')
+    save_name = os.path.join(SAVES_DIR, f"{AUTOSAVE_PREFIX if auto else ''}save-{sanitised_timestamp}.json")
+    with open(save_name, "w", encoding="utf-8") as save_file:
+        # We use chain.from_iterable() here because the quads array is 2D.
+        save = {
+            "quads": list(chain.from_iterable(game_state.board.quads)),
+            "players": game_state.players,
+            "heathens": game_state.heathens,
+            "turn": game_state.turn,
+            "cfg": game_state.board.game_config,
+            "night_status": {"until": game_state.until_night, "remaining": game_state.nighttime_left}
+        }
+        # Note that we use the SaveEncoder here for custom encoding for some classes.
+        save_file.write(json.dumps(save, cls=SaveEncoder))
+    save_file.close()
+
+
+def load_game(game_state: GameState, game_controller: GameController):
+    """
+    Loads the game with the given index from the saves/ directory.
+    :param save_idx: The index of the save file to load. Determined from the list of saves chosen from on the menu.
+    """
+    # Reset the namer so that we have our original set of names again.
+    game_controller.namer.reset()
+    # Sort and reverse both the autosaves and manual saves, remembering that the (up to) 3 autosaves will be
+    # displayed first in the list.
+    autosaves = list(filter(lambda file_name: file_name.startswith(AUTOSAVE_PREFIX), os.listdir(SAVES_DIR)))
+    saves = list(
+        filter(lambda file_name: not file_name == "README.md" and not file_name.startswith(AUTOSAVE_PREFIX),
+               [f for f in os.listdir(SAVES_DIR) if not f.startswith('.')]))
+    autosaves.sort()
+    autosaves.reverse()
+    saves.sort()
+    saves.reverse()
+    all_saves = autosaves + saves
+
+    try:
+        with open(os.path.join(SAVES_DIR, all_saves[game_controller.menu.save_idx]), "r",
+                  encoding="utf-8") as save_file:
+            # Use a custom object hook when loading the JSON so that the resulting objects have attribute access.
+            save = json.loads(save_file.read(), object_hook=ObjectConverter)
+            # Load in the quads.
+            quads = [[None] * 100 for _ in range(90)]
+            for i in range(90):
+                for j in range(100):
+                    quads[i][j] = save.quads[i * 100 + j]
+                    # The biomes require special loading.
+                    quads[i][j].biome = Biome[quads[i][j].biome]
+            game_state.players = save.players
+            # The list of tuples that is quads_seen needs special loading, as do a few other of the same type,
+            # because tuples do not exist in JSON, so they are represented as arrays, which will clearly not work.
+            for i in range(len(game_state.players[0].quads_seen)):
+                game_state.players[0].quads_seen[i] = (
```

Comment:
This looks like some new special logic, I don't think you can do this sort of [] access with sets, though.

Reply:
Hmm, the logic here is pretty much the same as how it was in the original `game.py`, e.g.

```
                    self.players[0].quads_seen[i] = (self.players[0].quads_seen[i][0], self.players[0].quads_seen[i][1])
```

Or have I missed something?

Reply:
Yeah turns out PyCharm is just wrong, at this point in the code the seen quads are still a `list` from the save file, and are converted to a set just below, so they'll have index access.